src/lib.rs

```diff
@@ -285,179 +302,31 @@ fn run_commands<C: Command>(runner: &Runner<C>) -> CommandHandle {
-            .spawn()
-            .unwrap_or_else(|_| panic!("Unable to spawn process: {}", command.command.clone()));
-        let std_out = cmd_handle.stdout.take();
-        let std_err = cmd_handle.stderr.take();
-        let mut std_out_handle = None;
-        let mut std_err_handle = None;
-
-        let shared_handle = Arc::new(Mutex::new(cmd_handle));
-
-        let child_clone = shared_handle.clone();
-        let kill_trigger_clone = kill_trigger.clone();
-        thread::spawn(move || kill_thread(&kill_trigger_clone, child_clone));
-
-        if let Some(output) = std_out {
-            let mut buffered_stdout = BufReader::new(output);
-            let new_name = command_name.clone();
-            let new_chan = send_chan.clone();
-            std_out_handle = Some(thread::spawn(move || {
-                read_stream(&new_name, new_chan, &mut buffered_stdout, true);
-            }));
-        }
-
-        if let Some(output) = std_err {
-            let mut buffered_stdout = BufReader::new(output);
-            let new_name = command_name.clone();
-            let new_chan = send_chan.clone();
-            std_err_handle = Some(thread::spawn(move || {
-                read_stream(&new_name, new_chan, &mut buffered_stdout, false);
-            }));
-        }
-
-        if let Some(handle) = std_out_handle {
-            let _ = handle.join();
-        }
-
-        if let Some(handle) = std_err_handle {
-            let _ = handle.join();
-        }
-
-        let exit_status = shared_handle.lock().unwrap().wait();
-        match exit_status {
-            Ok(status) => {
-                let _ = send_chan.send(OutputMessage {
-                    name: command_name.clone(),
-                    message: OutputMessagePayload::Done(status.code()),
-                });
-
-                match options.restart {
-                    RestartOptions::Continue => {
-                        return (command_name, Some(status));
-                    }
-                    RestartOptions::Restart => {
-                        if status.success() {
-                            return (command_name, Some(status));
-                        }
-                    }
-                    RestartOptions::Kill => {
-                        if !status.success() {
-                            let _ = kill_trigger.initiate_kill();
-                        }
-                        return (command_name, Some(status));
-                    }
-                };
-            }
-            Err(e) => {
-                let _ = send_chan.send(OutputMessage {
-                    name: command_name.clone(),
-                    message: OutputMessagePayload::Error(e),
-                });
-                return (command_name, None);
-            }
-        }
-    })
+fn check_command(exec_name: &str) -> Result<(), CommandError> {
+    which::which(exec_name).map_err(|_| CommandError::CommandNotFound(exec_name.to_string()))?;
+    Ok(())
```

Comment:
Thoughts on writing this as a `match` statement? I'm not sure myself, mostly trying to get your take here. There are meaningful operations in both arms: return an error or return an empty tuple.

Reply:
sure